codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -378,16 +384,34 @@ protected Http2FrameWriter frameWriter() {
      * Writes a {@code RST_STREAM} frame to the remote endpoint and updates the connection state appropriately.
      */
     @Override
-    public ChannelFuture writeRstStream(ChannelHandlerContext ctx, int streamId, long errorCode,
-            ChannelPromise promise) {
-        Http2Stream stream = connection().stream(streamId);
+    public ChannelFuture writeRstStream(final ChannelHandlerContext ctx, int streamId, long errorCode,
+            final ChannelPromise promise) {
+        final Http2Stream stream = connection().stream(streamId);
+        if (stream == null) {
+            // Don't write a RST_STREAM frame if we are not aware of the stream.
+            return promise.setFailure(streamError(streamId, INTERNAL_ERROR,
+                    "Stream %d not found while writing RST_STREAM frame", streamId));
+        }
+        if (stream.isResetSent()) {
+            // Don't write a RST_STREAM frame if we already wrote a RST_STREAM frame
+            return promise.setFailure(streamError(streamId, INTERNAL_ERROR,
```

Comment:
BTW I'm in favor of succeeding here.  RST_STREAM can happen for a lot of reasons and two parts of the code may not be aware of each other. I think from an API perspective it should just succeed, so long as only one goes out on the wire.  WDYT?